src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/AnonymousTypeManager.Templates.cs

```diff
@@ -163,60 +156,190 @@ private void CheckSourceLocationSeen(AnonymousTypePublicSymbol anonymous)
                 this.System_Object,
                 Compilation.GetSpecialType(SpecialType.System_IntPtr),
                 returnsVoid ? Compilation.GetSpecialType(SpecialType.System_Void) : null,
                 parameterCount,
                 refKinds);
-            return this.SynthesizedDelegates.GetOrAdd(key, new SynthesizedDelegateValue(this, synthesizedDelegate)).Delegate;
+            return this.SynthesizedDelegates.GetOrAdd(key, synthesizedDelegate);
         }
 
         private NamedTypeSymbol ConstructAnonymousDelegateImplementationSymbol(AnonymousDelegatePublicSymbol anonymous, int generation)
         {
             var typeDescr = anonymous.TypeDescriptor;
             Debug.Assert(typeDescr.Location.IsInSource); // AnonymousDelegateTemplateSymbol requires a location in source for ordering.
 
-            var fields = typeDescr.Fields;
-            bool returnsVoid = fields.Last().Type.IsVoidType();
-            int nTypeArguments = fields.Length - (returnsVoid ? 1 : 0);
-            var refKinds = default(RefKindVector);
-            if (fields.Any(f => f.RefKind != RefKind.None))
+            // If all parameter types and return type are valid type arguments, construct
+            // the delegate type from a generic template. Otherwise, use a non-generic template.
+            if (!allValidTypeArguments(typeDescr))
             {
-                refKinds = RefKindVector.Create(nTypeArguments);
+                var fields = typeDescr.Fields;
+                bool returnsVoid = fields.Last().Type.IsVoidType();
+                int nTypeArguments = fields.Length - (returnsVoid ? 1 : 0);
+                var refKinds = default(RefKindVector);
+                if (fields.Any(f => f.RefKind != RefKind.None))
+                {
+                    refKinds = RefKindVector.Create(nTypeArguments);
+                    for (int i = 0; i < nTypeArguments; i++)
+                    {
+                        refKinds[i] = fields[i].RefKind;
+                    }
+                }
+                var typeArgumentsBuilder = ArrayBuilder<TypeWithAnnotations>.GetInstance(nTypeArguments);
                 for (int i = 0; i < nTypeArguments; i++)
                 {
-                    refKinds[i] = fields[i].RefKind;
+                    typeArgumentsBuilder.Add(fields[i].TypeWithAnnotations);
+                }
+                var typeArguments = typeArgumentsBuilder.ToImmutableAndFree();
+                var template = SynthesizeDelegate(parameterCount: fields.Length - 1, refKinds, returnsVoid, generation);
+
+                Debug.Assert(typeArguments.Length == template.TypeParameters.Length);
+                return typeArguments.Length == 0 ?
+                    template :
+                    template.Construct(typeArguments);
+            }
+            else
+            {
+                var typeParameters = GetReferencedTypeParameters(typeDescr);
+                var key = getTemplateKey(typeDescr, typeParameters);
```

Comment:
>var key = getTemplateKey(typeDescr, typeParameters); [](http://example.com/codeflow?start=16&length=52)

Are we testing that we are unifying template for the same shape delegates from different locations?

Reply:
Including across different generic contexts.

Reply:
`SynthesizedDelegateTypes_25`, `_33`, `_34`, and others.